### DISCUSSION TITLE
How does the test code call the production code?

### DISCUSSION THREAD

**Discussion by mosofsky** [SamZiegler/vitest-cucumber-plugin#27] [General]
In the example of how to write a test, I don't see how the test code calls the production code:  

```
When('I ask whether it\'s Friday yet', function (state) {
    return _.set('answer',(state.today === 'Friday') ? 'TGIF' : 'Nope',state);
});
```
[source](https://github.com/samuel-ziegler/vitest-cucumber-plugin/blob/c7d4af6129f5b8a2c358ced812f57bd20441f4aa/README.md?plain=1#L104)

By contrast, the Cucumber example does show how the production code, namely `isItFriday`, is called:

```
When('I ask whether it\'s Friday yet', function () {
  this.actualAnswer = isItFriday(this.today);
});
```
[source](https://github.com/samuel-ziegler/vitest-cucumber-plugin/blob/c7d4af6129f5b8a2c358ced812f57bd20441f4aa/README.md?plain=1#L81)

The idea of passing the `state` object through makes sense in so much as the Given/When/Then statements successively alter it like in the classic party game Telephone.  But how does the `state` object relate to the production code?  In particular, if the production code is the function `isItFriday`, how does the When statement altering a state object entail calling `isItFriday`?

**Comment by samuel-ziegler** [comment]
It is, admittedly, not a good example.  I created issue #11 so that I remember to create a better one.  The state object has no relation to the production code.  It's for communicating information between steps.  In redux parlance, I roughtly think about the Gherkin step definitions as the actions and the JS implementations as the reducers.

In this example, let's say that isItFriday is the production code and rewrite it like this:
```
import { isItFriday } from '../../src/example.js';
import { Given, When, Then } from 'vitest-cucumber-plugin';
import _ from 'lodash/fp.js';
import { expect } from 'vitest'

Given('today is Sunday', function () {
    return { today : 'Sunday' };
});

Given('today is Friday', function () {
    return { today : 'Friday' };
});

When('I ask whether it\'s Friday yet', function (state) {
    return _.set('answer',isItFriday(state.today),state);
});

Then('I should be told {string}', function (state,[ answer ]) {
    expect(state.answer).toBe(answer);
    return state;
});
```

Here are the logs so that you can see the state transitions:
```
[12:04:04.715] INFO (7837): Given('today is Sunday') ({},[],null) => {"today":"Sunday"}
[12:04:04.718] INFO (7837): When('I ask whether it's Friday yet') ({"today":"Sunday"},[],null) => {"today":"Sunday","answer":"Nope"}
[12:04:04.722] INFO (7837): Then('I should be told {string}') ({"today":"Sunday","answer":"Nope"},["Nope"],null) => {"today":"Sunday","answer":"Nope"}
[12:04:04.723] INFO (7837): Given('today is Friday') ({"today":"Sunday","answer":"Nope"},[],null) => {"today":"Friday"}
[12:04:04.723] INFO (7837): When('I ask whether it's Friday yet') ({"today":"Friday"},[],null) => {"today":"Friday","answer":"TGIF"}
[12:04:04.724] INFO (7837): Then('I should be told {string}') ({"today":"Friday","answer":"TGIF"},["TGIF"],null) => {"today":"Friday","answer":"TGIF"}

```
Does that make more sense?

**Comment by mosofsky** [comment]
Hi Sam,

Ok that makes sense!

Passing the state object through the steps seems like a great idea!

Michael
________________________________
